### ISSUE TITLE
[Feature request] Dynamic or fixed resolution

### ISSUE THREAD

**Issue by marcospampi** [IrisShaders/Iris#757]
**Is your feature request related to a problem? Please describe.**
Currently, Minecraft's render resolution is tied to window/screen size, and not configurable.
Newer laptops have denser displays, yet they do often have low power iGPUs and little battery capacity and those 1080p pixels aren't that worthy of computing. A dynamic resolution or fixed resolution that reduce power usage and improve battery life would be a great addition for Minecraft Java Edition. 

**Describe the solution you'd like**
Iris should implement a dynamic or fixed resolution system:
- Fixed resolution by far is the most straightforward here, Iris gets configured to render the game at a certain resolution, but not the UI.
- Dynamic resolution is the trickiest, there are two points to look up on mobile devices: power usage and poopness of iGPUs. Ya know how dynamic resolution works, if a target FPS isn't reached, lower the resolution until it reaches the target FPS or a lower bound, this *should work* as mobile SoCs do throttle and give feedback by how bad the game is running.

**Describe alternatives you've considered**
Set the whole system screen res lower LoL.

**Additional context**
None really, the feature request addresses common day to day issues with laptops and low power computers in general.
Oh, my laptop don't reach 60fps with shaders and 1080p res without a dangerous boost during a 46°C summer.

**Comment by KyleUltimate** [comment]
For the fixed one, you could use resolution control+

**Comment by coderbot16** [comment]
While this is already sort of addressed by Resolution Control, I'd like a feature like this in Iris itself since it is very useful when using shaders even on high end systems. Even though I have a powerful GPU, I don't like when it heats all the way up to 85C and prefer to keep it at around 60% load or below, and reducing the resolution is a good way to do that. Maybe I can make my own version, or bundle resolution control.

**Comment by ThinkWithPbody** [comment]
Resolution Control has not been updated to 1.18, is there a chance that this feature can be implemented into iris?

**Comment by TheNukeLore** [comment]
Resolution control works perfectly fine on 1.18, and I’m pretty sure it’s unlikely the quality options will be added to iris

**Comment by SlimySpeck** [comment]
This suggestion sure be perfect for people who are on low end pc

**Comment by 44lr** [comment]
Would be nice to have this without depending on external mods, i used to force the rendering scale to something higher than 1.5 on optifine, to avoid the need for anti-aliasing / TAA
it made things look way crispier

**Comment by coderbot16** [comment]
I would just like to say that I am still aware of this, but I haven't had any time to get around to implementing any of it. I think Iris would definitely benefit from having general options for enabling FXAA, changing the render resolution, enabling upscaling driven by things like FSR, and more.

If someone else would like to investigate implementing this feature, please let me know. I'd welcome any PRs that implement these kinds of things.

**Comment by devopsdani** [comment]
I aggree, it'd be nice if we could change render quality, turning on/off specular and normal maps just like in Optifine.

**Comment by coderbot16** [comment]
ResolutionControl would be a good place to look as far as how a render quality feature would be implemented, though one source of complexity is that in OptiFine the resolution slider changes how the final pass works and it causes the shader to do the upscaling, while in ResolutionControl the scaling happens elsewhere.

I'm not sure which approach is best, since some packs do have custom scaling behavior with OptiFine's way, but many other packs break or look really bad with OptiFine's way.

**Comment by marcospampi** [comment]
I ain't skilled at all in Java, though sure could fork and code some spaghetti, but I'm kinda lost on where to look for in sauce code, some direction could help.
If not mistaken, should draw the world onto a scaled surface, then draw that surface scaled back with wanted filter, then UI can be drawn... simple talks, hard work lmao

**Comment by Lolothepro** [comment]
It could be logical to add the "performance" label.

**Comment by Lec312** [comment]
With the recent death of Resolution Control 3, which was the latest fork of RC there is no way to use either supersampling or downsampling functionality on Iris/Sodium with any of the latest versions. (QDAA is a very messy solution as well as DSR/VSR)

Optifine parity isn't a goal, but this kind of feature would is sorely needed in a game like MC where supersampling benefits high-end users and the same with low-end.

I hope this kind of feature isn't left to forks of mods, but built into the mod itself. 

Thank you

**Comment by HarmonyFriends** [comment]
> With the recent death of Resolution Control 3, which was the latest fork of RC there is no way to use either supersampling or downsampling functionality on Iris/Sodium with any of the latest versions. (QDAA is a very messy solution as well as DSR/VSR)
> 
> Optifine parity isn't a goal, but this kind of feature would is sorely needed in a game like MC where supersampling benefits high-end users and the same with low-end.
> 
> I hope this kind of feature isn't left to forks of mods, but built into the mod itself.
> 
> Thank you

Yeah, the sudden deletion of Resolution Control 3 has been incredibly frustrating and my laptop has gone from getting almost 120fps to below 60. I really think this is a feature that shouldn't be left to unreliable and untrustworthy external mods.

**Comment by Ozzuneoj** [comment]
> > With the recent death of Resolution Control 3, which was the latest fork of RC there is no way to use either supersampling or downsampling functionality on Iris/Sodium with any of the latest versions. (QDAA is a very messy solution as well as DSR/VSR)
> > Optifine parity isn't a goal, but this kind of feature would is sorely needed in a game like MC where supersampling benefits high-end users and the same with low-end.
> > I hope this kind of feature isn't left to forks of mods, but built into the mod itself.
> > Thank you
> 
> Yeah, the sudden deletion of Resolution Control 3 has been incredibly frustrating and my laptop has gone from getting almost 120fps to below 60. I really think this is a feature that shouldn't be left to unreliable and untrustworthy external mods.

I'm honestly amazed that it's nearly 2025 and no major mods for MC have built in support for the super-sampling\down-sampling that Optifine has done with the "Render Quality" setting for 10+ years. 

How is this not a thing? Why are we still chasing around forks of tiny projects that people have no time to maintain just to get proper full-frame anti-aliasing?

Frankly, it's ridiculous that the vanilla game with no mods doesn't already have the equivalent of Optifine's Render Quality setting (and the dynamic lights from torches for that matter). 

Iris or even Sodium would both be great places for this to be implemented, since they are available on NeoForge and Fabric and are nearly ubiquitous.

Personally, I need something for NeoForge... but the fact that nothing is even available for Fabric right now just boggles my mind.

What even happened to Render Control 3? I have never seen a mod so thoroughly wiped from the internet. Only a few scattered records of it exist and they are mainly at sketchy\scammy mod sites that don't have permission to distribute mods.

**Comment by Zolo101** [comment]
> Personally, I need something for NeoForge... but the fact that nothing is even available for Fabric right now just boggles my mind.

I've made a fork of ResolutionControl++ for Forge 1.20.1, NeoForge & Fabric 1.21.4 ([RenderScale Modrinth](https://modrinth.com/mod/renderscale)) to hold us over for a couple more months...

I don't mind attempting a PR for fixed resolution down the line, but I think this is better suited for Sodium, because then you could use it with Nvidium, etc. Idk, what do you guys think

**Comment by SomeoneIsWorking** [comment]
> I've made a fork of ResolutionControl++ for Forge 1.20.1, NeoForge & Fabric 1.21.4 ([RenderScale Modrinth]

This just makes the game smaller for me, if I set 0.5, it takes 1/4 screen on bottom left corner and the rest of the screen is black with render artifacts
Tried disabling every other mod that has anything to do with rendering, didn't help.
